Rules/CheckDotNetTypes.cs

```diff
@@ -0,0 +1,176 @@
+
+            if (ast == null) throw new ArgumentNullException(Strings.NullAstErrorMessage);
+            IEnumerable<Ast> funcs = ast.FindAll(testAst => testAst is FunctionDefinitionAst, true);
+            foreach (FunctionDefinitionAst func in funcs)
+            {
+                IEnumerable<Ast> memberExpressions = func.FindAll(testAst => testAst is MemberExpressionAst, true);
+
+                
+
+                //Get the type of the variable before checking if the type has the underliend methods/properties
+                string path = Path.GetDirectoryName(Assembly.GetExecutingAssembly().Location);
+                ReflectionTypeAnalysis.Instance.GetTypeFromAssembly(path +"../../Engine/ReferenceAssembly");
+                typeMethod = ReflectionTypeAnalysis.Instance.typeMethod;
+                typeProperty = ReflectionTypeAnalysis.Instance.typeProperties;
+                typeFields = ReflectionTypeAnalysis.Instance.typeFields;
+
+                foreach (MemberExpressionAst member in memberExpressions)
+                {
+                    if (member.Expression is VariableExpressionAst)
+                    {
+                        VariableExpressionAst varAst = member.Expression as VariableExpressionAst;
+
+                        //    //Get the type of the variable before checking if the type has the underliend methods/properties
+                        string type = Helper.Instance.GetVariableTypeFromAnalysis(varAst, func);
+                        string field = member.Member.Extent.Text;
+
+                        if (!typeMethod.ContainsKey(type))
+                        {
+                            if (
+                                !string.IsNullOrEmpty(type) &&
+                                !string.Equals("Microsoft.Windows.Powershell.ScriptAnalyzer.Unreached", type,
+                                    StringComparison.OrdinalIgnoreCase) &&
+                                !string.Equals("Microsoft.Windows.Powershell.ScriptAnalyzer.Undetermined", type,
+                                    StringComparison.OrdinalIgnoreCase) &&
+                                !type.Contains("[]"))
+                            {
+                                yield return
+                                    new DiagnosticRecord(
+                                        string.Format(CultureInfo.CurrentCulture, Strings.DotNetTypeUnavailableError,
+                                            type),
+                                        varAst.Extent, GetName(), DiagnosticSeverity.Warning, fileName);
+                            }
+                        }
+                        else
+                        {
+                            if (!typeMethod[type].Contains(field, StringComparer.OrdinalIgnoreCase) && !typeProperty[type].Contains(field, StringComparer.OrdinalIgnoreCase)
+                                && !typeFields[type].Contains(field,StringComparer.OrdinalIgnoreCase))
```

Comment:
We didn't check that typeFields doesn't have the key type so an exception may be thrown here.